### PR TITLE
support reuse and do not reuse connection for bolt

### PR DESCRIPTION
**Pull request by leizhiyuan**
### Motivation:

Explain the context, and why you're making that change.
To make others understand what is the problem you're trying to solve.

### Modification:

Describe the idea and modifications you've done.

### Result:

Fixes #<GitHub issue number>. 

If there is no issue then describe the changes introduced by this PR.

### CODE REVIEW

**Comment by leizhiyuan** [comment]
@ujjboy @JervyShi  please review the design of the feature, if almost ok, I will keep doing some extra work.

File: extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientTransport.java [hunk 1]
```diff
@@ -435,15 +416,19 @@ public void handleRpcRequest(SofaRequest request) {
 
     @Override
     public InetSocketAddress remoteAddress() {
+        Connection connection = connectionManager.getConnection(RPC_CLIENT, transportConfig, url);
```
Comment (JervyShi) [inline extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientTransport.java]: All like this line can use `fetchConnection()` instead.

Comment (leizhiyuan) [reply extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientTransport.java]: > All like this line can use `fetchConnection()` instead.

fixed

File: extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/ReuseBoltClientConnectionManager.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            if (currentCount <= 0) {
+                // 此长连接无任何引用，可以销毁
+                connectionRefCounter.remove(connection);
+                needDestroy = true;
+            } else {
+                needDestroy = false;
+            }
+        }
+        if (needDestroy) {
+            rpcClient.closeConnection(url);
```
Comment (JervyShi) [inline extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/ReuseBoltClientConnectionManager.java]: This operate is dangerous, before `rpcClient.closeConnection(url)` invoke, may have new ref get this connection in concurrent scene.

Comment (leizhiyuan) [reply extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/ReuseBoltClientConnectionManager.java]: fixed

File: extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientConnectionManager.java [hunk 3]
```diff
@@ -18,177 +18,39 @@
...
+ * @version $Id: BoltClientConnectionManager.java, v 0.1 2019年01月29日 11:58 bystander Exp $
  */
-class BoltClientConnectionManager {
-
-    /**
-     * slf4j Logger for this class
-     */
-    private final static Logger                            LOGGER               = LoggerFactory
-                                                                                    .getLogger(BoltClientConnectionManager.class);
+public interface BoltClientConnectionManager {
```
Comment (ujjboy) [inline extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientConnectionManager.java]: it' better to change to an abstract class, and move common logic such as `addHook ` to this class.

File: extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/AloneBoltClientConnectionManager.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import com.alipay.sofa.rpc.base.Destroyable;
+import com.alipay.sofa.rpc.common.annotation.VisibleForTesting;
+import com.alipay.sofa.rpc.context.RpcRuntimeContext;
+import com.alipay.sofa.rpc.core.exception.SofaRpcRuntimeException;
+import com.alipay.sofa.rpc.log.Logger;
+import com.alipay.sofa.rpc.log.LoggerFactory;
+import com.alipay.sofa.rpc.transport.ClientTransportConfig;
+
+/**
+ * @author <a href="mailto:zhanggeng.zg@antfin.com">GengZhang</a>
```
[the remark this answers is not in the document]
Comment (leizhiyuan) [reply extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/AloneBoltClientConnectionManager.java]: changed

File: extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientConnectionManager.java (hunk 3 above)
Comment (leizhiyuan) [reply extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/BoltClientConnectionManager.java]: ok

**Pull request merged**